### DISCUSSION TITLE
How do I go about changing the max controlnet models settings through the API?

### DISCUSSION THREAD

**Discussion by monis24** [Mikubill/sd-webui-controlnet#722] [Q&A]
I'm trying to change the multi controlnet max models setting through the api, but can't seem to figure out how. I tried using override_settings with "setting_control_net_max_models_num" as the key, but I got a keyerror. Any idea how I go about doing this correctly?

Was able to find the setting using the GET /config route

<img width="636" alt="Screenshot 2023-04-09 at 7 18 49 PM" src="https://user-images.githubusercontent.com/52933919/230813625-ca0d3b69-cb75-49d2-aa68-9bd40919c1f1.png">

**Comment by ljleb** [comment]
There is no way to do what you want to do at the moment. The code on the side of the extension is written so that you can specify as many as you like, so in theory this would be unnecessary, only a limitation in the user interface.

In some situations you can use as many controlnet models as you wish. However at the moment a limitation in the implementation of the webui api makes it impossible to specify a different number of units than the max. See https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/9130 for a fix.

**Reply by monis24** [reply]
Ahh alright for sure. I've been trying to just use 2 models with multi controlnet like this 

```
"alwayson_scripts": {
            "ControlNet": {
                "args": [{
                    "input_image": cannyImage,
                    "module":"none",
                    "model": cannyModel,
                    "weight": 0.3,
                    "resize_mode": "Scale to Fit (Inner Fit)",
                    "guidance": 1,
                    "guidance_start": 0,
                    "guidance_end": 1,
                    "guessmode": false},
                    {
                    "input_image": poseImage,
                    "module":"none",
                    "model": poseModel,
                    "weight": 0.4,
                    "resize_mode": "Scale to Fit (Inner Fit)",
                    "guidance": 1,
                    "guidance_start": 0,
                    "guidance_end": 1,
                    "guessmode": false}
                ]
            }
        }
```

But it's only been using the first canny model, and not the second pose model. Do you see anything wrong?

**Reply by ljleb** [reply]
Apart from the usage of strings instead of ids for resize mode (check the docs for less error prone int values), it seems like it should work. If your max models is set to 1 that may be why it does not work. You could try raising the value of `"control_net_max_models_num"` in your `config.json` if you need a programming solution, although this needs a reload of the config IIUC and I am not sure there is a way from the api to do that (although I could be wrong?).

**Reply by monis24** [reply]
Thanks for the help, I'll try some stuff out

**Reply by monis24** [reply]
> Apart from the usage of strings instead of ids for resize mode (check the docs for less error prone int values), it seems like it should work. If your max models is set to 1 that may be why it does not work. You could try raising the value of `"control_net_max_models_num"` in your `config.json` if you need a programming solution, although this needs a reload of the config IIUC and I am not sure there is a way from the api to do that (although I could be wrong?).

Do you think it would be possible to hard code the value in? I can't seem to find where in the repo this variable is being interacted with

**Reply by ljleb** [reply]
Yes you can add the value in `config.json` at the root of the webui directory, or update its value if it is already there. The ui uses it here

https://github.com/Mikubill/sd-webui-controlnet/blob/187ae88038af6f4daa91d5dc941564d9a4df90ef/scripts/controlnet.py#L421

the webui reads the amount of gradio components returned from this function to determine the number of elements expected in the `alwayson_scripts` parameter of the rest api routes.

**Reply by ljleb** [reply]
If you need to change this value programatically, you could do something like this I presume: (not tested)

```python
shared.opts.set("control_net_max_models_num", 4)
shared.opts.save(shared.config_filename)
```

**Reply by monis24** [reply]
So i tried to hard code it with the possible programmatic solution you provided. I tried placing the two lines at 428 & 429, and also 888 & 889, and got a key error both times for "control_net_max_models_num", honestly not sure how I could do it another way.

**Reply by monis24** [reply]
Ideally I want to be able to hard code it so I can spin up instances of A1111 and be able to use multi controlnet without setting the max model to 2 and then having to restart my instance.

**Reply by kft334** [reply]
Try to set control_net_max_models_num via the POST /sdapi/v1/options API route if you haven't yet.
